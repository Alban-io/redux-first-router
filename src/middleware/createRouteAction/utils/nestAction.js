import typeToScene from '../../../utils/typeToScene'

export default (action, { prev: p, ...prev }, history, basename) => {
  const { kind, entries, index, length, location } = history
  const { url, pathname, search } = location
  const { type, payload = {}, query = {}, state = {}, hash = '' } = action
  const scene = typeToScene(type)

  return {
    ...action,
    type,
    payload,
    query,
    state,
    hash,
    location: {
      ...action.location,
      url,
      pathname,
      search,
      basename,
      scene,

      prev,

      kind,
      entries,
      index,
      length
    }
  }
}
