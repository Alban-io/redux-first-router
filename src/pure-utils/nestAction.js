// @flow
import type { Action, Location, ReceivedAction, History } from '../flow-types'


export default (
  pathname: string,
  receivedAction: ReceivedAction,
  prev: Location,
  history: History,
  isMiddleware: boolean,
  kind?: string,
): Action => {
  const { type, payload = {}, meta } = receivedAction
  const shouldPushEntry = isMiddleware && history.entries[history.index].pathname !== pathname

  const entries = shouldPushEntry
    ? history.entries.slice(0, history.index + 1).map(entry => entry.pathname)
    : history.entries.map(entry => entry.pathname)

  return {
    type,
    payload,
    meta: {
      ...meta,
      location: {
        current: {
          pathname,
          type,
          payload,
        },
        prev,
        load: kind === 'load' ? true : undefined,
        backNext: kind === 'backNext' ? true : undefined,
        history: {
          entries: shouldPushEntry ? [...entries, pathname] : entries,
          index: shouldPushEntry ? history.index + 1: history.index,
          length: shouldPushEntry ? history.index + 2 : history.length,
        },
      },
    },
  }
}
