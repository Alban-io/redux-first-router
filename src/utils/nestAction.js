export default function nestAction(pathname, receivedAction, prev) {
  let {type, payload, meta} = receivedAction;
  
  return {
    type, 
    payload,
    meta, 
    location: {
      current: {
        pathname,
        type, 
        payload,
      },
      prev,
    }
  };
}