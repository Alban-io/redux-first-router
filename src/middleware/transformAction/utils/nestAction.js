import { typeToScene, isNotFound } from '../../../utils'

export default (action, prevState, history, fromAction) => {
  const { kind, entries, index, length, location } = history
  const { pathname, search, basename: bn, url: u, state = {} } = location
  const { type, params = {}, query = {}, hash = '' } = action
  const prev = createPrev(prevState)
  const from = createFrom(fromAction)
  const scene = typeToScene(type)
  const basename = bn.substr(1)
  const url = bn + u
  const status = kind === 'redirect'
    ? ((action.location && action.location.status) || 302)
    : (isNotFound(type) ? 404 : 200)

  return {
    ...action,
    type,
    params,
    query,
    hash,
    state,
    basename,
    location: {
      ...action.location,
      url,
      pathname,
      search,
      scene,
      status,

      prev,
      from,

      kind,
      entries,
      index,
      length
    }
  }
}

const createPrev = (prevState) => {
  const prev = { ...prevState }

  delete prev.prev
  delete prev.hasSSR
  delete prev.from
  delete prev.location

  return prev
}

const createFrom = (fromAction) => {
  if (!fromAction) return null

  // if redirect(action) from outside of pipeline, we receive the state instead (see ./reduxAction.js)
  if (!fromAction.location) {
    const from = { ...fromAction }

    delete from.prev
    delete from.hasSSR
    delete from.from

    return from
  }

  // if redirect occurred during pipeline, we receive an action representing the previous state
  const { type, params, query, state, hash, basename, location } = fromAction
  const from = { type, params, query, state, hash, basename, ...location }

  delete from.prev
  delete from.hasSSR
  delete from.from

  return from
}
