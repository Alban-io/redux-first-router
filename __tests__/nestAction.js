import { createMemoryHistory, createBrowserHistory } from 'rudy-history'
import nestAction, { nestHistory } from '../src/pure-utils/nestAction'

it('nestAction properly formats/nests action object', () => {
  const history = createMemoryHistory()
  const pathname = '/path'
  const receivedAction = {
    type: 'FOO',
    payload: { bar: 'baz' },
    meta: { info: 'something' }
  }
  const location = {
    pathname: 'previous',
    type: 'PREV',
    payload: { bla: 'prev' }
  }

  let action = nestAction(
    pathname,
    receivedAction,
    location,
    history
  ) /*? $.meta */

  expect(action.type).toEqual('FOO')
  expect(action.payload).toEqual({ bar: 'baz' })

  expect(action.type).toEqual(action.meta.location.current.type)
  expect(action.payload).toEqual(action.meta.location.current.payload)

  expect(action.meta.location.prev).toEqual(location)
  expect(action.meta).toMatchObject(receivedAction.meta)
  expect(action.meta.location.current.pathname).toEqual(pathname)

  expect(action).toMatchSnapshot()

  expect(action.meta.location.kind).not.toBeDefined()
  action = nestAction(pathname, receivedAction, location, history, 'load')
  expect(action.meta.location.kind).toEqual('load')

  action = nestAction(pathname, receivedAction, location, history, 'pop')
  expect(action.meta.location.kind).toEqual('pop')
})

it('nestHistory formats simplified history object for action + state', () => {
  const history = createMemoryHistory() // still use `createMemoryHistory` for stability during tests
  history.push('/foo')
  history.push('/bar/baz')

  const nestedHistory = nestHistory(history) /*? */
  expect(nestedHistory).toMatchSnapshot()
})

it('nestHistory returns undefined when using createBrowserHistory', () => {
  const history = createBrowserHistory()
  const nestedHistory = nestHistory(history)
  expect(nestedHistory).toEqual(undefined)
})
