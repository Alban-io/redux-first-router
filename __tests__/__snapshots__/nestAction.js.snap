// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nestAction properly formats/nests action object 1`] = `
Object {
  "kind": undefined,
  "meta": Object {
    "info": "something",
    "location": Object {
      "basename": undefined,
      "current": Object {
        "index": 0,
        "kind": undefined,
        "length": 1,
        "pathname": "/path",
        "payload": Object {
          "bar": "baz",
        },
        "type": "FOO",
      },
      "history": Object {
        "entries": Array [
          Object {
            "hash": "",
            "key": "345678",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "index": 0,
        "kind": undefined,
        "length": 1,
      },
      "kind": undefined,
      "prev": Object {
        "pathname": "previous",
        "payload": Object {
          "bla": "prev",
        },
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "bar": "baz",
  },
  "type": "FOO",
}
`;
